cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,29 +285,406 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+			devStr := partName
+			if unlockRes.Device != "" {
+				devStr += fmt.Sprintf(" (device %s)", unlockRes.Device)
+			}
+			m.degradedState.LogErrorf("cannot unlock encrypted %s with sealed run key: %v", devStr, err)
+			part.UnlockState = partitionErrUnlocking
+
+		} else {
+			// TODO: we don't know if this is a plain not found or  a different error
+			m.degradedState.LogErrorf("cannot locate %s partition for mounting host data: %v", part, err)
+		}
+
+		return nil
+	}
+
+	if unlockRes.IsDecryptedDevice {
+		part.UnlockState = partitionUnlocked
+		part.UnlockKey = keyRun
+	}
+
+	return nil
+}
+
+func (m *stateMachine) setUnlockStateWithFallbackKey(partName string, unlockRes secboot.UnlockResult, err error) error {
+	part := m.degradedState.partition(partName)
+
+	// first check the result and error for consistency; since we are using udev
+	// there could be inconsistent results at different points in time
+	// TODO: when we refactor UnlockVolumeUsingSealedKeyIfEncrypted to not also
+	//       find the partition on the disk, we should eliminate this
+	//       consistency checking as we can code it such that we don't get these
+	//       possible inconsistencies
+	// ensure consistency between encrypted state of the device/disk and what we
+	// may have seen previously
+	if m.isEncryptedDev && !unlockRes.IsDecryptedDevice {
+		// then we previously were able to positively identify an
+		// ubuntu-data-enc but can't anymore, so we have inconsistent results
+		// from inspecting the disk which is suspicious and we should fail
+		return fmt.Errorf("inconsistent disk encryption status: previous access resulted in encrypted, but now is unencrypted from partition %s", partName)
+	}
+
+	// if isEncryptedDev hasn't been set on the state machine yet, then set that
+	// on the state machine before continuing - this is okay because we might
+	// not have been able to do anything with ubuntu-data if we couldn't mount
+	// ubuntu-boot, so this might be the first time we tried to unlock
+	// ubuntu-data and m.isEncryptedDev may have the default value of false
+	if !m.isEncryptedDev && unlockRes.IsDecryptedDevice {
+		m.isEncryptedDev = unlockRes.IsDecryptedDevice
+	}
+
+	// also make sure that if we previously saw a device that we see the same
+	// device again
+	if unlockRes.Device != "" && part.Device != "" && unlockRes.Device != part.Device {
+		return fmt.Errorf("inconsistent partitions found for %s: previously found %s but now found %s", partName, part.Device, unlockRes.Device)
+	}
+
+	if unlockRes.Device != "" {
+		part.FindState = partitionFound
+		part.Device = unlockRes.Device
+	}
+
+	if !unlockRes.IsDecryptedDevice && unlockRes.Device != "" && err != nil {
+		// this case should be impossible to enter, if we have an unencrypted
+		// device and we know what the device is then what is the error?
+		return fmt.Errorf("internal error: inconsistent return values from UnlockVolumeUsingSealedKeyIfEncrypted for partition %s", partName)
```

Comment:
Should we include the "err" we got here in the returned error string?

Reply:
done in https://github.com/snapcore/snapd/pull/9621